arrow2_convert_derive/src/derive_enum.rs

```diff
@@ -0,0 +1,539 @@
+
+                    Box::new(arrow2::array::UnionArray::from_data(
+                        <#original_name as arrow2_convert::field::ArrowField>::data_type().clone(),
+                        std::mem::take(&mut self.types).into(),
+                        values,
+                        #offsets_take
+                    ))
+                }
+
+                fn as_arc(&mut self) -> std::sync::Arc<dyn arrow2::array::Array> {
+                    let values = vec![#(
+                        <#mutable_variant_array_types as arrow2::array::MutableArray>::as_arc(&mut self.#variant_names),
+                    )*];
+
+                    std::sync::Arc::new(arrow2::array::UnionArray::from_data(
+                        <#original_name as arrow2_convert::field::ArrowField>::data_type().clone(),
+                        std::mem::take(&mut self.types).into(),
+                        values,
+                        #offsets_take
+                    ))
+                }
+
+                fn as_any(&self) -> &dyn std::any::Any {
+                    self
+                }
+
+                fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
+                    self
+                }
+
+                fn push_null(&mut self) {
+                    use arrow2::array::TryPush;
+                    self.try_push(None::<#original_name>).unwrap();
+                }
+
+                fn shrink_to_fit(&mut self) {
+                    #(
+                        <#mutable_variant_array_types as arrow2::array::MutableArray>::shrink_to_fit(&mut self.#variant_names);
+                    )*
+                    self.types.shrink_to_fit();
+                    #offsets_shrink_to_fit
+                }
+            }
+        };
+
+        let field_arrow_serialize_impl = quote! {
+            impl arrow2_convert::serialize::ArrowSerialize for #original_name {
+                type MutableArrayType = #mutable_array_name;
+
+                #[inline]
+                fn new_array() -> Self::MutableArrayType {
+                    Self::MutableArrayType::default()
+                }
+
+                #[inline]
+                fn arrow_serialize(v: &Self, array: &mut Self::MutableArrayType) -> arrow2::error::Result<()> {
+                    use arrow2::array::TryPush;
+                    array.try_push(Some(v))
+                }
+            }
+        };
+
+        generated.extend([
+            array_decl,
+            array_impl,
+            array_arrow_mutable_array_impl,
+            array_try_push_impl,
+            array_default_impl,
+            array_try_extend_impl,
+            array_mutable_array_impl,
+            field_arrow_serialize_impl,
+        ])
+    }
+
+    if gen_deserialize {
+        let array_name = &input.common.array_name();
+        let iterator_name = &input.common.iterator_name();
+
+        // - For dense unions, return the value of the variant that corresponds to the matched arm. Since
```

Comment:
@jorgecarleitao This is logic going to change once scalar support comes in so that we can serialize slices correctly.